src/util/exception_utils.h

```diff
@@ -33,4 +33,23 @@ class invalid_user_input_exceptiont
   std::string what() const noexcept;
 };
 
+class system_exceptiont
+{
+private:
+  std::string reason;
+
+public:
+  system_exceptiont(const std::string &reason) : reason(reason)
```

Comment:
Should be marked as explicit, should take reason as plain string rather than reference and should `move` reason.